### PR TITLE
Add telemetry for repository usage

### PR DESCRIPTION
**Pull request by DaveCTurner**
Adds to the `GET _cluster/stats` endpoint information about the snapshot
repositories in use, including their types, whether they are read-only
or read-write, and for Azure repositories the kind of credentials in
use.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Closes ES-9070

File: server/src/main/java/org/elasticsearch/action/admin/cluster/stats/RepositoryUsageStats.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Map;
+
+public record RepositoryUsageStats(Map<String, Map<String, Long>> statsByType) implements Writeable, ToXContentObject {
```
Comment (pxsalehi) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/stats/RepositoryUsageStats.java:20]: could you please document what the map(s) represent?

Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/stats/RepositoryUsageStats.java:20]: do we need to have some serialization tests for this or is it covered somewhere else?

File: docs/reference/cluster/stats.asciidoc [hunk 2]
```diff
@@ -1579,6 +1579,9 @@ The API returns the following response:
    },
    "snapshots": {
      ...
+   },
+   "repositories": {
```
Comment (pxsalehi) [inline docs/reference/cluster/stats.asciidoc:1583]: is there any other documentation, other than adding the key here?

**Review by pxsalehi — COMMENTED** [review]
Some questions and suggestions.

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java [hunk 3]
```diff
@@ -366,4 +374,8 @@ private String deriveURIFromSettings(boolean isPrimary) {
             throw new IllegalArgumentException(e);
         }
     }
+
+    public Set<String> usageFeatures() {
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java]: Added Javadocs in 61566556319b

File: server/src/main/java/org/elasticsearch/action/admin/cluster/stats/RepositoryUsageStats.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/stats/RepositoryUsageStats.java:25]: Added Javadocs in 61566556319b.

File: docs/reference/cluster/stats.asciidoc (hunk 2 above)
Comment (DaveCTurner) [reply docs/reference/cluster/stats.asciidoc:1605]: Added a little more docs in 61566556319b.

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java [hunk 4]
```diff
@@ -150,6 +152,12 @@ private AzureStorageSettings(
         this.endpointSuffix = endpointSuffix;
         this.timeout = timeout;
         this.maxRetries = maxRetries;
+        this.usageFeatures = Strings.hasText(key) ? Set.of("uses_key_credentials")
+            : Strings.hasText(sasToken) ? Set.of("uses_sas_token")
+            : SocketAccess.doPrivilegedException(() -> System.getenv("AZURE_FEDERATED_TOKEN_FILE")) == null
+                ? Set.of("uses_default_credentials", "uses_managed_identity")
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java:158]: My thinking is that a repository _is_ `read-only` so no verb needed, but if we just called the features `sas_token` etc. then that doesn't match. But honestly I don't care, if you feel strongly enough I'll go through and change this in all the places where it appears in tests etc.

**Review by pxsalehi — APPROVED** [review]
LGTM!

**Pull request merged**